pkg/agent/cniserver/ipam/antrea_ipam.go

```diff
@@ -147,29 +147,36 @@ func (d *AntreaIPAM) Add(args *invoke.Args, k8sArgs *types.K8sArgs, networkConfi
 	}
 
 	owner := getAllocationOwner(args, k8sArgs, reservedOwner, false)
-	var ip net.IP
-	var subnetInfo *crdv1a2.SubnetInfo
-	if reservedOwner != nil {
-		ip, subnetInfo, err = allocator.AllocateReservedOrNext(crdv1a2.IPAddressPhaseAllocated, owner)
-	} else if len(ips) == 0 {
-		ip, subnetInfo, err = allocator.AllocateNext(crdv1a2.IPAddressPhaseAllocated, owner)
-	} else {
-		ip = ips[0]
-		subnetInfo, err = allocator.AllocateIP(ip, crdv1a2.IPAddressPhaseAllocated, owner)
-	}
-	if err != nil {
-		return true, nil, err
-	}
+	result := IPAMResult{Result: current.Result{CNIVersion: current.ImplementedSpecVersion}}
+	for i := range allocators {
+		var ip net.IP
+		var subnetInfo *crdv1a2.SubnetInfo
+		if reservedOwner != nil {
+			ip, subnetInfo, err = allocators[i].AllocateReservedOrNext(crdv1a2.IPAddressPhaseAllocated, owner)
+		} else if len(ips) == 0 {
+			ip, subnetInfo, err = allocators[i].AllocateNext(crdv1a2.IPAddressPhaseAllocated, owner)
+		} else {
+			ip = ips[i]
+			subnetInfo, err = allocators[i].AllocateIP(ip, crdv1a2.IPAddressPhaseAllocated, owner)
+		}
+		if err != nil {
+			return true, nil, err
```

Comment:
Where do we release the allocated IPs if one allocator already successfully allocated IPs before the failure?

Reply:
Nice catch. Added rollback logic here.